rfcs/0042-config-option.md

```diff
@@ -0,0 +1,191 @@
+        for documentation on the supported values.
+      '';
+    };
+  };
+  
+  config = mkIf cfg.enable {
+  
+    # Set minimal config to get service working by default
+    services.foo.config = {
+      dataPath = "/var/lib/foo";
+      logLevel = mkDefault "WARN";
+    };
+  
+    environment.etc."foo.json".text = configText;
+    
+    networking.firewall.allowedTCPPorts = [ port ];
+    # ...
+  };
+}
+```
+
+This approach solves all* of the above mentioned problems. In addition we have the following properties that work with this approach out of the box:
+- Ability to easily query arbitrary configuration values with `nix-instantiate --eval '<nixpkgs/nixos>' -A config.services.foo.config`
+- The configuration file is well formatted with the right amount of indentation everywhere
+- Usually hardcoded defaults can now be replaced by simple assignment of the `config` option, which also allows people to override those values with `mkForce`
+
+*: The only problem it doesn't solve is the coupling to upstream defaults for options that need to be known at evaluation time, such as `port` in the example above, but that's not really avoidable.
+
+### Configuration types
+
+A set of types for common configuration formats should be provided in `lib.types.config`. Such a type should encode what values can be set in files of this configuration format as a Nix value, with the module system being able to merge multiple values correctly. This is the part that checks whether the user set an encodeable value. This can be extended over time, but could include the following as a start:
+- JSON
+- YAML, which is probably the same as JSON
+- INI
+- A simple `key=value` format
+- A recursive `key.subkey.subsubkey=value` format
+
+Sometimes programs have their own configuration formats, in which case the type should be implemented in the program's module directly.
+
+### Configuration format writers
+
+To convert the Nix value into the configuration string, a set of configuration format writers should be provided under `lib.config`. These should make sure that the resulting text is somewhat properly formatted with readable indentation. These writers will have to include ones for all of the above-mentioned configuration types. As with the type, if the program has its own configuration format, the writer should be implemented in its module directly.
+
+### Documentation
+
+The nixpkgs manual should be updated to recommend this way of doing program configuration in modules, along with examples.
+
+## Limitations
+
+- Limited to configuration file formats representable conveniently in Nix, such as JSON, YAML, INI, key-value files, or similar formats. Examples of unsuitable configuration formats are Haskell, Lisp, Lua or other generic programming languages. For those it is recommended to not hardcode anything and provide a `config` option with type `types.str` or `types.lines` if it makes sense to merge multiple assignments of it.
```

Comment:
I think this very much depends on the program's configuration format. For example it would be more confusing than helpful to add some semi-hardcoded default string to xmonad's Haskell configuration format, I'd rather just have a `config` option for that without any specialized NixOS options.

And with Bash it makes both sense to have semi-hardcoded defaults (stuff like `HISTFILE=${cfg.histFile}`) which you can always override later on anyways, and a `types.lines` option with which you can add arbitrary commands.

The only reason I can think of that a bunch of NixOS options, *without* a `types.lines` option are reasonable is when the program is configured with a programming language, when it wouldn't actually have needed to be a programming language, because the configuration is so simple.

Reply:
I think there are many cases where for most users most of the Turing-complete config is just about simple assignments which could be handled in the style of this RFC, but advanced users add many things that need `types.lines` because they use the actual programming I expect this to hold for `bashrc`, for example.

So the optimal choice between free form config and structured values config might include having both…

Reply:
Basic configurations in such environments usually don't need the full complexity (even my initial XMonad config didn't need the power of Haskell when I used it for my desktop). I guess it's best to allow the module's author to decide what's the best fit for such a case - several dedicated options or an `extraConfig` option.

Reply:
My point is that we might end up having key-value config _and_ `types.lines` addition without manual description of options.

Reply:
@7c6f434c I guess it won't be called `config` then though, but e.g. for bash env vars it would be `programs.bash.environment`. And it also wouldn't be a special type like json, but just `attrsOf str` or so.

But yeah, sometimes it makes sense to have both specialized options and `types.lines`, I think this should be left to the author in general.

I'll add a section on that.

Reply:
I expanded that section with what we discussed here.